src/pd/client.rs

```diff
@@ -108,6 +108,58 @@ pub trait PdClient: Send + Sync + 'static {
         })
         .boxed()
     }
+
+    // Returns a Steam which iterates over the contexts for ranges in the same region.
+    fn group_ranges_by_region(
+        self: Arc<Self>,
+        mut ranges: Vec<BoundRange>,
+    ) -> BoxStream<'static, Result<(RegionId, Vec<BoundRange>)>> {
+        ranges.reverse();
```

Comment:
I think there is another 'group by regions' function, can it share code with this new one?

Reply:
Do you mean group `group_keys_by_region` and `group_ranges_by_region` together into `group_by_region`?

Reply:
Yeah, if it is possible and not too ugly.

Reply:
I tried this by implementing a `IntoRange` trait for key data types.
```rust
pub trait IntoRange {
    fn into_keys(self) -> (Key, Option<Key>);
    fn set_start_key(&mut self, k: Key);
    fn set_end_key(&mut self, k: Key);
}
```
After implementing this trait for `Key` and `BoundRange` I found some key data types are imported from other packages like `kvproto::protos::kvrpcpb::KvPair` and `kvproto::protos::kvrpcpb::Mutation` which makes the this work more strange. Besides, how we group ranges is more complex than group keys, because we need to split ranges and push the right part back into the range vector in this process, the logic can be redundant for key like data.
I don't have good ideas about how to doing this, what your opinion?

Reply:
Yeah, seems like that is more work that it is worth. Maybe it is possible to factor out a helper function or two, or a macro? Otherwise don't worry about it.

Reply:
Let me try out macro~